File: src/tq/engine/planner.py
"""Run planning for target-scoped tq engine execution."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from tq.discovery.filesystem import build_analysis_index
from tq.engine.context import AnalysisContext

if TYPE_CHECKING:
    from tq.config.models import TqTargetConfig


@dataclass(frozen=True, slots=True)
class PlannedTargetRun:
    """Planned execution unit for a target-scoped analysis run."""

    target: TqTargetConfig
    context: AnalysisContext


def plan_target_runs(
    *,
    configured_targets: tuple[TqTargetConfig, ...],
    active_targets: tuple[TqTargetConfig, ...],
) -> tuple[PlannedTargetRun, ...]:
    """Plan deterministic target runs before rule execution."""
    known_target_package_paths = tuple(
        configured_target.package_path.as_posix()
        for configured_target in configured_targets
    )

    planned_runs: list[PlannedTargetRun] = []
    for target in active_targets:
        index = build_analysis_index(
            source_root=target.source_package_root,
            test_root=target.test_root,
        )
        context = AnalysisContext.create(
            index=index,
            settings={
                "target_name": target.name,
                "package_path": target.package_path.as_posix(),
                "known_target_package_paths": known_target_package_paths,
                "test_root_display": target.test_root.name,
            },
        )
        planned_runs.append(PlannedTargetRun(target=target, context=context))

    return tuple(planned_runs)

File: crates/tq-engine/src/planner.rs
use tq_discovery::build_analysis_index;

use crate::{AnalysisContext, EngineError, PlannedTargetRun, TargetContext, TargetPlanInput};

pub fn plan_target_runs(
    configured_targets: &[TargetPlanInput],
    active_targets: &[TargetPlanInput],
) -> Result<Vec<PlannedTargetRun>, EngineError> {
    let known_target_package_paths = configured_targets
        .iter()
        .map(|target| target.package_path().clone())
        .collect::<Vec<_>>();

    let mut planned_runs = Vec::with_capacity(active_targets.len());
    for target in active_targets {
        let index = build_analysis_index(target.source_package_root(), target.test_root())
            .map_err(EngineError::Discovery)?;

        let target_context = TargetContext::new(
            target.name().clone(),
            target.package_path().clone(),
            known_target_package_paths.clone(),
            target.test_root_display().to_path_buf(),
        );

        planned_runs.push(PlannedTargetRun::new(
            target.clone(),
            AnalysisContext::with_target(index, target_context),
        ));
    }

    Ok(planned_runs)
}
